### ISSUE TITLE
general question about "no test files" and coverage

### ISSUE THREAD

**Issue by jspiro** [mattn/goveralls#84]
Coverage reports don't include directories that have no tests. Let's say there are two .go files in it:

```
$ go test -cover lib/missing
?   	lib/missing	[no test files]
```

If we add a single `_test.go` file to that package that does nothing, coverage is suddenly reported for all statements in all files in the package (of course, all zeroes).

How can we be sure that a coverage report reflects coverage of ALL packages in a tree, not just those that have at least one test? You think you have such a great coverage until you realize most of your code has no coverage because there are no tests.

One solution is CI failing when a package is discovered with no tests, so you add one. A better solution would be getting coverage output for these packages. Annoyingly, `go test` can't be forced to generate a report, and it doesn't return a non-zero error code when there are no tests, so it's unnecessarily hard (like all things in Go...) to automate the detection of packages without tests.

What do people generally do in this situation?

**Comment by mattn** [comment]
goveralls upload coverage information from coverprofile which is generated by `go test -coverprofile=xxx`. when there is no tests, coverprofile is not generated. So goveralls doesn't upload anything. Maybe travis CI return error on the test.

**Comment by jspiro** [comment]
Right! But doesn't that feel wrong, if you have 100 packages of Go code, and only one has tests, and those tests cover 100% of that package's code, then coverage for the repo will be 100%.

**Comment by retr0h** [comment]
> Right! But doesn't that feel wrong, if you have 100 packages of Go code, and only one has tests, and those tests cover 100% of that package's code, then coverage for the repo will be 100%.

I like this behavior b/c I have a test utility package I use only in tests as helpers.  I do not want these covered, and they should be skipped since a test file does not exist.

I'd just add something to your CI to force it to fail.

```
[[ $(go test -cover ./... | grep 'no test files') ]] && false
```
